source/opt/loop_unroller.cpp

```diff
@@ -0,0 +1,553 @@
+
+  // We need to account for the initial body when calculating the remainder.
+  uint32_t remainder = loop->NumIterations() % (factor + 1);
+
+  // Construct the constant.
+  uint32_t resultId = ir_context_->TakeNextId();
+  ir::Operand constant(spv_operand_type_t::SPV_OPERAND_TYPE_LITERAL_INTEGER,
+                       {remainder});
+  std::unique_ptr<ir::Instruction> newConstant(new ir::Instruction(
+      ir_context_, SpvOp::SpvOpConstant, uint32_type_id, resultId, {constant}));
+
+  uint32_t constant_id = newConstant->result_id();
+  ir_context_->module()->AddGlobalValue(std::move(newConstant));
+
+  // Add the merge block to the back of the binary.
+  blocks_to_add_.push_back(
+      std::unique_ptr<ir::BasicBlock>(new_loop.GetMergeBlock()));
+
+  // Reset the usedef analysis.
+  ir_context_->InvalidateAnalysesExceptFor(
+      ir::IRContext::Analysis::kAnalysisNone);
+  opt::analysis::DefUseManager* def_use_manager =
+      ir_context_->get_def_use_mgr();
+
+  // Add the blocks to the function.
+  AddBlocksToFunction(loop->GetMergeBlock());
+
+  // Update the condition check.
+  ir::Instruction& conditional_branch = *loop->GetConditionBlock()->tail();
+  ir::Instruction* condition_check =
+      def_use_manager->GetDef(conditional_branch.GetSingleWordOperand(0));
+  condition_check->SetInOperand(1, {constant_id});
+
+  // Update the next phi node.
+  ir::Instruction* new_induction = new_loop.GetInductionVariable();
+  new_induction->SetInOperand(0, {constant_id});
+  new_induction->SetInOperand(1, {new_merge_id});
+
+  ir_context_->InvalidateAnalysesExceptFor(
+      ir::IRContext::Analysis::kAnalysisNone);
+
+  ir_context_->ReplaceAllUsesWith(loop->GetMergeBlock()->id(), new_merge_id);
+}
+
+// Duplicate the |loop| body |factor| number of times while keeping the loop
+// backedge intact.
+void LoopUnrollerUtilsImpl::PartiallyUnroll(ir::Loop* loop, int factor) {
+  Unroll(loop, factor);
+  AddBlocksToFunction(loop->GetMergeBlock());
+}
+
+// Duplicate the |loop| body |factor| number of times while keeping the loop
+// backedge intact.
+void LoopUnrollerUtilsImpl::Unroll(ir::Loop* loop, int factor) {
+  state_ = LoopStateUtil{loop};
+  for (int i = 0; i < factor; ++i) {
+    CopyBody(loop, true);
+  }
+
+  // The first condition block is perserved until now so it can be copied.
+  FoldConditionBlock(loop->GetConditionBlock(), 1);
+
+  uint32_t phi_index = 5;
+  uint32_t phi_variable =
+      state_.previous_phi_->GetSingleWordOperand(phi_index - 1);
+  uint32_t phi_label = state_.previous_phi_->GetSingleWordOperand(phi_index);
+
+  ir::Instruction* original_phi = loop->GetInductionVariable();
+
+  // SetInOperands are offset by two.
+  // TODO: Work out why.
+  original_phi->SetInOperand(phi_index - 3, {phi_variable});
+  original_phi->SetInOperand(phi_index - 2, {phi_label});
+}
+
+void LoopUnrollerUtilsImpl::FullyUnroll(ir::Loop* loop) {
+  Unroll(loop, loop->NumIterations() - 1);
```

Comment:
`loop->NumIterations() - 1` because you reuse the loop body to do the last iter ? If so put a comment to say so, otherwise they might be an issue with the method